# npm-lockfile-38f99c3374ca4e9bd75f3ec34f3edb249eb391cf.zip

## 프로젝트 개요

- 분석 날짜: 2024-06-04 22:40:38
- 언어: JavaScript
- 프레임워크: NPM Library
- 코드 패치 LLM 모델: GPT-4o

## 취약점 분석 결과

### 0. 취약점 분석 통계   

- 전체 발견 취약점: 1
- 심각도별 취약점   
    - error: 0
    - warning: 1
    - note : 0

### 1. Indirect uncontrolled command line
- 코드 경로: `/npm-lockfile-38f99c3374ca4e9bd75f3ec34f3edb249eb391cf/getLockfile.js`
   
- 심각도: warning
   
- 설명: Forwarding command-line arguments to a child process executed within a shell may indirectly introduce command-line injection vulnerabilities.
   
- 메시지: `This command depends on an unsanitized [["command-line argument"|"relative:///npm-lockfile-38f99c3374ca4e9bd75f3ec34f3edb249eb391cf/bin.js:17:2:21:2"]].`
     
- 원본 코드   
```js
'use strict';

const path = require('path');
const { exec } = require('child_process');
const promisify = require('util.promisify');
const inspect = require('object-inspect');
const colors = require('colors/safe');
const copyFileCB = require('fs-copy-file');

const copyFile = promisify(copyFileCB);
const readFile = promisify(require('fs').readFile);

const getProjectTempDir = require('./getProjectTempDir');

module.exports = function getLockfile(packageFile, date, {
	npmNeeded,
	only,
	logger = () => {},
} = {}) {
	if (typeof packageFile !== 'string' || packageFile.length === 0) {
		return Promise.reject(colors.red(`\`packageFile\` must be a non-empty string; got ${inspect(packageFile)}`));
	}
	if (typeof date !== 'undefined' && !new Date(date).getTime()) {
		return Promise.reject(colors.red(`\`date\` must be a valid Date format if provided; got ${inspect(date)}`));
	}
	const tmpDirP = getProjectTempDir({ npmNeeded, logger });
	const npmRC = path.join(path.dirname(packageFile), '.npmrc');
	const copyPkg = tmpDirP.then((tmpDir) => {
		logger(colors.blue(`Creating \`package.json\` in temp dir for ${date || '“now”'} lockfile`));
		return Promise.all([
			copyFile(packageFile, path.join(tmpDir, 'package.json')),
			copyFile(npmRC, path.join(tmpDir, '.npmrc')).catch((err) => {
				if (!err || !(/^ENOENT: no such file or directory/).test(err.message)) {
					throw err;
				}
			}),
		]);
	});
	return Promise.all([tmpDirP, copyPkg]).then(([tmpDir]) => new Promise((resolve, reject) => {
		const PATH = path.join(tmpDir, '../node_modules/.bin');
		logger(colors.blue(`Running npm install to create lockfile for ${date || '“now”'}...`));
		exec(
			`npm install --ignore-scripts --package-lock --package-lock-only${date ? ` --before=${date}` : ''}${only ? ` --only=${only}` : ''}`,
			{
				cwd: tmpDir,
				env: {
					PATH: `${PATH}:${process.env.PATH}`,
					NODE_ENV: process.env.NODE_ENV,
				},
			},
			(err) => {
				if (err) {
					reject(err);
				} else {
					resolve(tmpDir);
				}
			}
		);
	})).then((tmpDir) => {
		logger(colors.blue(`Reading lockfile contents for ${date || '“now”'}...`));
		const lockfile = path.join(tmpDir, 'package-lock.json');
		return readFile(lockfile, { encoding: 'utf-8' });
	});
};

```
   
- 패치 코드   
```js
'use strict';

const path = require('path');
const { spawn } = require('child_process');
const promisify = require('util.promisify');
const inspect = require('object-inspect');
const colors = require('colors/safe');
const copyFileCB = require('fs-copy-file');

const copyFile = promisify(copyFileCB);
const readFile = promisify(require('fs').readFile);

const getProjectTempDir = require('./getProjectTempDir');

module.exports = function getLockfile(packageFile, date, {
	npmNeeded,
	only,
	logger = () => {},
} = {}) {
	if (typeof packageFile !== 'string' || packageFile.length === 0) {
		return Promise.reject(colors.red(`\`packageFile\` must be a non-empty string; got ${inspect(packageFile)}`));
	}
	if (typeof date !== 'undefined' && !new Date(date).getTime()) {
		return Promise.reject(colors.red(`\`date\` must be a valid Date format if provided; got ${inspect(date)}`));
	}
	const tmpDirP = getProjectTempDir({ npmNeeded, logger });
	const npmRC = path.join(path.dirname(packageFile), '.npmrc');
	const copyPkg = tmpDirP.then((tmpDir) => {
		logger(colors.blue(`Creating \`package.json\` in temp dir for ${date || '“now”'} lockfile`));
		return Promise.all([
			copyFile(packageFile, path.join(tmpDir, 'package.json')),
			copyFile(npmRC, path.join(tmpDir, '.npmrc')).catch((err) => {
				if (!err || !(/^ENOENT: no such file or directory/).test(err.message)) {
					throw err;
				}
			}),
		]);
	});
	return Promise.all([tmpDirP, copyPkg]).then(([tmpDir]) => new Promise((resolve, reject) => {
		const PATH = path.join(tmpDir, '../node_modules/.bin');
		logger(colors.blue(`Running npm install to create lockfile for ${date || '“now”'}...`));

        let args = ['install', '--ignore-scripts', '--package-lock', '--package-lock-only'];
        if (date) args.push(`--before=${date}`);
        if (only) args.push(`--only=${only}`);

        const process = spawn('npm', args, {
            cwd: tmpDir,
            env: {
                PATH: `${PATH}:${process.env.PATH}`,
                NODE_ENV: process.env.NODE_ENV,
            },
        });

        process.on('close', (code) => {
            if (code !== 0) {
                reject(new Error(`npm process exited with code ${code}`));
            } else {
                resolve(tmpDir);
            }
        });
    })).then((tmpDir) => {
		logger(colors.blue(`Reading lockfile contents for ${date || '“now”'}...`));
		const lockfile = path.join(tmpDir, 'package-lock.json');
		return readFile(lockfile, { encoding: 'utf-8' });
	});
};

```
   
- 코드 비교   
```diff
--- 

+++ 

@@ -1,7 +1,7 @@

 'use strict';
 
 const path = require('path');
-const { exec } = require('child_process');
+const { spawn } = require('child_process');
 const promisify = require('util.promisify');
 const inspect = require('object-inspect');
 const colors = require('colors/safe');
@@ -39,24 +39,27 @@

 	return Promise.all([tmpDirP, copyPkg]).then(([tmpDir]) => new Promise((resolve, reject) => {
 		const PATH = path.join(tmpDir, '../node_modules/.bin');
 		logger(colors.blue(`Running npm install to create lockfile for ${date || '“now”'}...`));
-		exec(
-			`npm install --ignore-scripts --package-lock --package-lock-only${date ? ` --before=${date}` : ''}${only ? ` --only=${only}` : ''}`,
-			{
-				cwd: tmpDir,
-				env: {
-					PATH: `${PATH}:${process.env.PATH}`,
-					NODE_ENV: process.env.NODE_ENV,
-				},
-			},
-			(err) => {
-				if (err) {
-					reject(err);
-				} else {
-					resolve(tmpDir);
-				}
-			}
-		);
-	})).then((tmpDir) => {
+
+        let args = ['install', '--ignore-scripts', '--package-lock', '--package-lock-only'];
+        if (date) args.push(`--before=${date}`);
+        if (only) args.push(`--only=${only}`);
+
+        const process = spawn('npm', args, {
+            cwd: tmpDir,
+            env: {
+                PATH: `${PATH}:${process.env.PATH}`,
+                NODE_ENV: process.env.NODE_ENV,
+            },
+        });
+
+        process.on('close', (code) => {
+            if (code !== 0) {
+                reject(new Error(`npm process exited with code ${code}`));
+            } else {
+                resolve(tmpDir);
+            }
+        });
+    })).then((tmpDir) => {
 		logger(colors.blue(`Reading lockfile contents for ${date || '“now”'}...`));
 		const lockfile = path.join(tmpDir, 'package-lock.json');
 		return readFile(lockfile, { encoding: 'utf-8' });
```
   
- 코드 패치 설명  
    - patched